source/common/router/router.cc

```diff
@@ -1130,12 +1151,33 @@ void Filter::UpstreamRequest::encodeTrailers(const Http::HeaderMap& trailers) {
   if (!request_encoder_) {
     ENVOY_STREAM_LOG(trace, "buffering trailers", *parent_.callbacks_);
   } else {
+    // Encodes metadata if exists.
+    if (!downstream_metadata_map_vector_.empty()) {
+      ENVOY_STREAM_LOG(trace, "Send metadata before sending trailers. {}", *parent_.callbacks_,
+                       downstream_metadata_map_vector_);
+      request_encoder_->encodeMetadata(downstream_metadata_map_vector_);
+      downstream_metadata_map_vector_.clear();
+    }
+
     ENVOY_STREAM_LOG(trace, "proxying trailers", *parent_.callbacks_);
     request_encoder_->encodeTrailers(trailers);
     upstream_timing_.onLastUpstreamTxByteSent(parent_.callbacks_->dispatcher().timeSource());
   }
 }
 
+void Filter::UpstreamRequest::encodeMetadata(Http::MetadataMapPtr&& metadata_map_ptr) {
+  if (!request_encoder_) {
+    ENVOY_STREAM_LOG(trace, "request_encoder_ not ready. Store metadata_map to encode later: {}",
+                     *parent_.callbacks_, *metadata_map_ptr);
+    downstream_metadata_map_vector_.emplace_back(std::move(metadata_map_ptr));
+  } else {
+    ENVOY_STREAM_LOG(trace, "Encode metadata: {}", *parent_.callbacks_, *metadata_map_ptr);
+    Http::MetadataMapVector metadata_map_vector;
```

Comment:
Similar question about why we need to create a local vector just to encode the map?

Reply:
This is to convert from unique_ptr<MetadataMap> to MetadataMapVector (aka vector<unique_ptr<MetadataMap>>) to preserve the possible multiple copies of the same metadata. (encodeMetadata's input is MetadataMapVector).